### ISSUE TITLE
Fresh error with GAP in stable branch

### ISSUE THREAD

**Issue by alex-konovalov** [gap-packages/anupq#15]
```
########> Diff in /home/hudson/hudson/workspace/GAP-pkg-update-stable-quicktes\
t/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/standard/label/graupius/GAP-pkg-update\
-stable-snapshot/tst/teststandard/bugfix.tst, line 2977:
# Input is:
if LoadPackage("anupq",false) <> fail then
for i in [1..192] do Q:=Pq( FreeGroup(2) : Prime:=3, ClassBound:=1 ); od; fi;
# Expected output:
# But found:
Panic: cannot fork to subprocess.
Error, sorry! Run out of pseudo-ttys. Can't open an io stream.

########
```

**Comment by fingolfin** [comment]
I cannot reproduce this here. But then it really depends on what method it uses to obtain pseudo ttys. Can you perhaps attach the config.h of the machine with the error?

**Comment by fingolfin** [comment]
To clarify: the config.h is interesting because I would like to know which of HAVE_POSIX_OPENPT / HAVE_PTSNAME / HAVE_GETPT / HAVE_PTSNAME_R / ... are set; this determines which ptty method is used, and that might explain why it works on some systems and not on others

**Comment by alex-konovalov** [comment]
Yes - here it is: 
[config.h.txt](https://github.com/gap-packages/anupq/files/97736/config.h.txt)

**Comment by fingolfin** [comment]
OK, thanks to a test case @heikodietrich gave me, I was able to reproduce this. After investigating closer, I discovered that the message regarding pseudo-ttys is a red herring -- it is printed by anupq itself! This is the relevant code:
```
    iorec.stream := InputOutputLocalProcess( ANUPQData.tmpdir, 
                                             ANUPQData.binary, 
                                             opts );
    if iorec.stream = fail then
      Error( "sorry! Run out of pseudo-ttys. Can't open an io stream.\n" );
    fi;
```
And while `InputOutputLocalProcess` only returns fail `CREATE_PTY_IOSTREAM` returned fail, the actual cause of that failure can be something other than running out of pseudo-ttys. Indeed, in this case, the message before that is key: `Panic: cannot fork to subprocess`. This message is generated by the GAP kernel, by function `StartChildProcess` in `iostream.c`. 

So the fork failed. But why exactly is not clear, and GAP does not help, because it does not analyze the error condition further. At the very least, it should print POSIX `errno`. I'll modify my version of GAP to do so, and I'll also hack it to use `explain_fork`. Perhaps that will give some further insights.

**Comment by gregg0** [comment]
Dear  Max

This is not a red herring. When one tries to create an iostream, one can run out of the pseudottys.
It typically happens when one opens and closes iostreams continually.
Of course, it is possible  I might be able to recover a pseudotty once an iostream had closed.
Is there such an approach, that you know of?

Regards,
Greg
________________________________________

**Comment by fingolfin** [comment]
Greg,
this is a red herring in so far as that the message is misleading: it suggest that the only possible cause for the error is that the system runs out of pttys -- but in reality, this is only one of multiple possible causes for that error.

And yes, normally, pttys should always be automatically be recovered when the associated child process dies. There used to be a bug in GAP which prevented more than 256 (or so) child processes, but I fixed that many years ago. 

Note that GAP prints an error "open pseudo tty failed" if openng a pty fails, but this error was *not* triggered. Instead, GAP proceeded to fork(), and *that* failed...

But of course it's still possible that there is another bug related to ptys, which might be the source of this issue. But it could just as well be something completely else. After all, when debugging, it's important to keep an open mind, and not jump to hasty conclusions :-).

Anyway, as described, I restarted the computations with a modified version of GAP, which should provide more insights as to what causes fork to fail

**Comment by gregg0** [comment]
Thanks Max

When running tests back in 2001 or so, I noticed that sometimes I couldn't get an IOStream ... and I essentially responded with the message that GAP gave.
It seemed to make sense ... so I didn't take it any further. It also seemed to be a problem one could avoid by keeping an IOStream open, rather than repeatedly opening and closing them.
Presumably, Heiko ran into this problem in a way, which for him the above approach was not desirable.

I can see it would have been useful to have you around back in 2001.
I wanted to rewrite Eamonn's C program because I hated the way it could fail with a segmentation fault rather than respond with a diagnostic saying something like" "You can't do that because it will cause such and such to be exceeded".
Werner stopped me getting involved with such idea ... probably you would have too ... but then I get the feeling you may have modified the C code in places to cause the program to exit more gracefully.

Regards,
Greg
________________________________________

**Comment by fingolfin** [comment]
Yeah, I think a segfault is never acceptable -- the program should orderly exit with a sane error message :-).
As it is, I think it would be interesting to rewrite anupq with modern technology, perhaps using C++ or even a higher level language like Julia...

Anyway, that's getting quite off-topic :-). So far, my modified version of GAP did not trigger the error, but I keep trying, so hopefully we'll soon know more.

**Comment by fingolfin** [comment]
I finally reproduced the issue, with my extra GAP debugging code active. This is what I saw:
```
Panic: cannot fork to subprocess (errno 12, fork() failed, Cannot allocate
memory (12, ENOMEM) because insufficient kernel memory was available).
```
So, pseudo ttys are indeed not the culprit -- at least in my instance of the problem. And running out of kernel memory seems like a pretty good reason :-). 

Two immediate tasks that follow:
* improve the GAP error message to at least give errno)
* change the anupq code to print a better error message.
